sdk/storage/Azure.Storage.Common/src/Shared/ExpectContinueOnThrottlePolicy.cs

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Threading;
+using System;
+using Azure.Core;
+using Azure.Core.Pipeline;
+
+internal class ExpectContinueOnThrottlePolicy : HttpPipelineSynchronousPolicy
+{
+    private long _lastThrottleTicks = 0;
+
+    private long _backoffTicks = TimeSpan.TicksPerMinute;
+    public TimeSpan Backoff
+    {
+        get => TimeSpan.FromTicks(_backoffTicks);
+        set => _backoffTicks = value.Ticks;
+    }
+
+    public long ContentLengthThreshold { get; set; }
+
+    public override void OnSendingRequest(HttpMessage message)
+    {
+        if (message.Request.Content == null ||
+            (message.Request.Content.TryComputeLength(out long contentLength) && contentLength < ContentLengthThreshold))
+        {
+            return;
+        }
+        long lastThrottleTicks = Interlocked.Read(ref _lastThrottleTicks);
+        if (DateTimeOffset.UtcNow.Ticks - lastThrottleTicks < _backoffTicks)
+        {
+            message.Request.Headers.SetValue("Expect", "100-continue");
+        }
+    }
+
+    public override void OnReceivedResponse(HttpMessage message)
+    {
+        base.OnReceivedResponse(message);
+        if (message.HasResponse && (
+            message.Response.Status == 429 ||
+            message.Response.Status == 500 ||
+            message.Response.Status == 503))
```

Comment:
I think we also wanted to enable 100-continue on TCP RST? According to the email thread this would be `SocketException` with a [SocketError](https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socketerror?view=net-8.0) of `ConnectionReset`. Not sure how that would manifest in the pipeline message though.

Email also mentions "timeouts" but I'm a bit unclear on what type of timeout they mean. Likely network timeouts and/or HTTP 408?

Reply:
Yes I plan to add support for RST in a future PR. The main issue is that RST doesn't just happen in the pipeline, and so it can't be handled just there. While this policy needs to be what adds the header, we need a broader solution to have failures like that from elsewhere in the library signal the error has been encountered.